2주차 미션 제출

jambottle opened the pull request:
# 2주차 미션 제출 (김재원)

## 구현 화면

<img width="810" alt="week2_01" src="https://user-images.githubusercontent.com/72926450/149742337-f61e00d2-f48c-40e1-b917-70313bcc62ba.png">
<img width="810" alt="week2_02" src="https://user-images.githubusercontent.com/72926450/149742361-1e94fbde-1a3a-4d75-9018-b7576a1bd039.png">

## 접근 방법

* [x] 미션 코드를 구현하기 위해서 어떤 생각들을 했는지 정리해보세요.

**📌 주어진 요구사항을 정확히 숙지하고, 그에 따라 페이지를 개발하였습니다.**
- 페이지를 구성할 때에는 모바일 환경의 웹 페이지를 기준으로 개발 (* [Unsplash](https://unsplash.com/)에서 무료로 배포 중인 이미지들을 활용)
- ItemInfo.vue 파일에 2주차 미션으로 구현한 기능들을 모아 단일 컴포넌트로 페이지를 구성
- **[리팩토링]**
  - Item 데이터와 직접적인 관련성이 떨어지는 부분을 별도의 하위 컴포넌트로 분리 (Shop/Review)
  - 하위 컴포넌트를 ItemInfo 컴포넌트에 등록하고, props로 필요한 데이터만 선택적으로 전달
  - `$emit`을 활용해 Shop 컴포넌트에서 상위 컴포넌트의 이벤트를 실행 (toggleLike)

**📌 아직 Jest와 Vue Test Utils, TDD 방법론이 낯설지만, 최선을 다해 적용해보았습니다.**

> TDD 방법론의 3단계에 따라 **1)** 요구사항에 맞춰 테스트 코드를 작성하고, **2)** 작성한 모든 테스트가 성공할 때까지 최소한의 코드 수정을 먼저 수행한 뒤, **3)** 그 밖의 레이아웃 구현 및 스타일링을 덧붙이는 방식으로 코드를 작성함으로써 지난주에 부족했던 부분을 극복하려 노력했습니다.

- Jest와 Vue Test Utils를 활용하여 여섯 가지 미션 요구사항에 대한 테스트 코드를 먼저 작성
- 새로운 데이터 값을 가정했을 때의 테스트 결과를 확인하기 위해 setData 메서드 활용
- 앞서 작성한 테스트 코드를 바탕으로 ItemInfo 컴포넌트의 전반적인 레이아웃 구현
- **[리팩토링]**
  - 직접적인 관련성이 높은 데이터 세트를 기준으로 테스트 코드를 세 그룹으로 구분 (Shop/Item/Review)
  - 같은 데이터를 공유하는 테스트 그룹의 경우, beforeEach 메서드를 활용해 테스트마다 데이터 초기화

**📌 API가 마련되지 않은 상황이 가정되었기 때문에 프로토타입 구현에 Mock Data를 활용하였습니다.**
- 데이터베이스의 공백을 해결하기 위해 상품 상세정보(`itemData`)와 리뷰 목록(`reviewData`)을 Mock Data로 대체
- **[리팩토링]**
  - 각각의 Mock Data 내에서 관련성이 높은 상위 단계의 개념끼리 묶는 방향으로 데이터 구조를 보다 체계적으로 수정

**📌 시각적 요구사항 구현을 위해 Vue.js의 템플릿 문법, 조건부 렌더링 등의 기본 문법들을 적극 활용하였습니다.**
- v-for 디렉티브를 활용하여 판매자를 설명하는 해시태그 목록과 리뷰 목록의 반복을 구현
- v-if/else와 v-on 디렉티브의 조합으로 "좋아요" 아이콘 클릭에 따른 Toggle 효과 구현 (toggleLike)
- v-if/else 디렉티브를 활용하여 할인가와 정상 가격을 비교한 결과에 따른 조건부 렌더링 실시
- v-html 디렉티브를 활용하여 일반 텍스트로 해석될 수 있는 상품 설명을 HTML 형태로 변환 등

## 특이사항

* [x] 구현하면서 궁금했던 점이나, 아쉬웠던 부분이 있으면 적어주세요.

𝑸. 이번 주에는 특히 DOM Node의 존재 여부를 확인하는 테스트를 위주로 작성하게 되었습니다. 주어진 요구사항에는 스타일링이나 템플릿 문법과 관련된 내용도 포함되어 있었는데, 이에 대한 테스트도 필요했던 것인지 궁금하고, 추가로 어떤 테스트 코드를 더 작성할 수 있을지도 궁금합니다! 🤔

## 제출 후 확인

* [x] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.
* [x] Pull Request 생성 후 Reivewer에 @externship-master 와 @lkaybob 을 추가해주세요.

jambottle committed:
Chore: Font Awesome & node-sass, sass-loader 패키지 설치

jambottle committed:
Add: reset.scss & common.scss 추가 및 적용

jambottle committed:
Add: 상품정보 및 리뷰 Mock Data 작성 및 추가

jambottle committed:
Test: 2주차 미션 요구사항에 대한 컴포넌트 단위 테스트 코드 작성
- Jest와 Vue Test Utils를 활용하여 여섯 가지 미션 요구사항에 대한 테스트 코드 작성

jambottle committed:
Add: 2주차 미션 요구사항 구현
- 앞서 작성한 테스트 코드를 바탕으로 ItemInfo 컴포넌트의 전반적인 레이아웃 구현 및 스타일링
- 스타일링 시, CSS 스타일링 규칙만을 가지고 있는 W3.CSS 프레임워크를 main.js에 일괄 적용

jambottle committed:
Add: 2주차 미션 요구사항 추가 구현 - (1) 상품 좋아요 기능
- vue-fontawesome 패키지의 Vue 3 호환성 문제 해결 (3.0.0-5 버전으로 설치)
- 상품 상세정보 Mock Data에 좋아요 여부를 저장하는 itemLiked 항목 추가
- v-if/else와 v-on 디렉티브의 조합으로 아이콘 클릭에 따른 Toggle 효과 구현

jambottle committed:
Add: 2주차 미션 요구사항 추가 구현 - (2) 장바구니 구매 알림 기능
- 1주차 미션 때 구현해둔 Modal 컴포넌트를 재사용하여 장바구니 알림 기능 구현
- 장바구니 구매 버튼 클릭에 따라 장바구니에 담긴 상품명, 가격과 함께 알림창 등장
- 그 밖에도 CSS Nesting 단계 및 일부 스타일링 순서 수정/보완

jambottle committed:
Test: 2주차 미션 요구사항 추가 구현에 대한 컴포넌트 단위 테스트 실시
- Jest와 Vue Test Utils를 활용하여 자체 추가한 두 가지 기능에 대한 테스트 코드 작성
- 새로운 데이터 값을 가정했을 때의 테스트 결과를 확인하기 위해 setData 메서드 활용

jambottle committed:
Add: 2주차 미션 요구사항 추가 구현 - (3) 상품 할인 여부에 따른 조건부 렌더링 구현
- 상품의 할인가가 정상 가격과 같을 경우, 할인율 및 할인가를 표시하지 않는 기능 구현
- v-if/else 디렉티브를 활용하여 할인가와 정상 가격을 비교한 결과에 따른 렌더링 실시

jambottle committed:
Add: 2주차 미션 요구사항 추가 구현 - (4) 좋아요 기능 수정 및 리뷰 개수 표시
- 좋아요의 대상이 상품이 아닌 판매자임을 인지하고, Mock Data에 shopLiked 항목 추가
- 상품에 달린 리뷰의 개수를 리뷰 목록의 배열에 length 메서드를 이용하여 계산 후 표시

lkaybob commented on mission/src/assets/itemData.js:
```diff
@@ -0,0 +1,18 @@
+export default {
+  shopName: '브라운슈즈',
```
상위단계의 개념끼리 묶일 수 있었으면 더 좋았을 것 같습니다.

```
{
  shop: {
    name: ...
    tag: ...
    ...
  },
  item: {
    name: ...,
    ...
  }
}
```

jambottle replied:
말씀하신 대로 수정해보았습니다! 🙏

lkaybob commented on mission/src/main.js:
```diff
@@ -2,5 +2,6 @@ import { createApp } from 'vue';
 import App from './App.vue';
 import router from './router';
 import store from './store';
+import '../node_modules/w3-css/w3.css';
```
npm에서의 설명을 참고하시면 아래와 같이 import하실 수 있습니다.

https://www.npmjs.com/package/w3-css

```
import 'w3-css/4/w3pro.css';
```

jambottle replied:
말씀하신 대로 수정해보았습니다! 🙏

lkaybob commented on mission/src/views/ItemInfo.vue:
```diff
@@ -1,24 +1,306 @@
+      <p v-if="itemPrice.discount !== itemPrice.original">
+        <b data-test="discount-rate"> {{ Math.round(discountRate) }}% </b>
+        <span data-test="discount-price">
+          {{ itemPrice.discount.toLocaleString() }}원
```
inline expression은 되도록 이면 computed property로 정의하시는 것을 권장드립니다.

jambottle replied:
말씀하신 대로 수정해보았습니다! 🙏

lkaybob commented on mission/tests/unit/ItemInfo/ItemInfo.spec.js:
```diff
@@ -1,10 +1,60 @@
+    });
+    await wrapper.get('button[data-test="footer-button"]').trigger('click');
+
+    expect(wrapper.get('div[class="modal-body"]').text()).toContain('삼선 슬리퍼');
```
존재여부 뿐만 아니라 이런 식으로 테스트 데이터에 대한 data property를 주입했을 때 이를 잘 rendering하는지 테스트하시면 됩니다.

jambottle replied:
앞선 테스트 코드들을 작성하면서 아쉬움이 느껴져서 새로운 기능들의 추가와 함께 좀 다르게 테스트를 짜봐야겠다고 생각했습니다..!!

피드백 남겨주신 것처럼 `setData`를 통해 **주입한 Data Property가 잘 출력되는지**에 대해서도 테스트할 수 있도록
기존의 테스트 코드들도 다시 수정해봐야겠습니다. 매주 TDD 방법론에 대해 알아가고 조금씩 익숙해지고 있어 신기하고 재밌습니다 😁

jambottle replied:
지난 라이브 코딩을 복기해보면서 **Data Property를 주입하는 방식의 테스트 코드**도 몇 개 더 구현해보았습니다 🙌

Soohyeon02 commented:
안녕하세요 재원님! code presentation 잘 보았습니다!
재원님의 code presentation을 보면서 인상 깊었던 점은 재원님은 코드를 작성하실 때 현재 목표만이 아닌 추후에 다른 여러 상황들을 고려하며 효율적으로 작성하시는 것 같다는 생각이 들었습니다!!
초보자인 저는 재원님 코드를 보면서 매번 배워가고 있어요😮😮
이번주 미션 수행하느라 수고 많으셨습니다!

lkaybob reviewed (commented):
* template을 semantic한 tag들을 사용하여 구현해주신 점이 돋보입니다.
* 데이터 요구사항들을 도출해내어 이를 data property에 잘 정의해주셨습니다. 
* 다만 데이터 구조화가 덜 된 점이 아쉽습니다.
> 𝑸. 이번 주에는 특히 DOM Node의 존재 여부를 확인하는 테스트를 위주로 작성하게 되었습니다. 주어진 요구사항에는 스타일링이나 템플릿 문법과 관련된 내용도 포함되어 있었는데, 이에 대한 테스트도 필요했던 것인지 궁금하고, 추가로 어떤 테스트 코드를 더 작성할 수 있을지도 궁금합니다! 🤔

라인 코멘트를 참고해주세요.

Soohyeon02 commented on mission/src/views/ItemInfo.vue:
```diff
@@ -19,7 +19,14 @@
+        >
+          <FontAwesomeIcon v-if="itemLiked" :icon="icon.like" />
+          <FontAwesomeIcon v-else :icon="icon.unlike" />
+        </div>
```
저는 좋아요 기능은 구현하기가 어렵게 느껴져서 포기하고 이미지를 삽입하였는데 재원님은 v-if와 v-on을 사용하셔서 구현하셨군요. 배워갑니다!!

allegoryj commented:
안녕하세요 재원님! 퀄리티 높은 code presentation 잘 보았습니다:)
단일 컴포넌트로 페이지를 구현하셨음에도 시멘틱 태그들과 명시적인 이름들을 사용하여 각 구성 요소의 역할들을 이해하는데 수월하였습니다!!
개인적으로 mock data를 사용해서 데이터들을 불러오신 점이 가장 인상깊었습니다.
세심한 부분 ( ex. 소수점 올림, 가격 쉼표 표시 등) 까지 놓치지 않고 페이지를 구현하신 점이 크게 배울 부분인 것 같습니다!!
이번 주도 고생많으셨습니다!

watertheqlant commented on mission/src/App.vue:
```diff
@@ -3,6 +3,9 @@
 </template>
 
 <style>
+@import './assets/scss/reset.scss';
+@import './assets/scss/common.scss';
```
css파일 분리를 하셨군요! 
공통 스타일을 쓰거나 유지보수성에도 좋겠네요 :)

allegoryj replied:
우와아.. 저도 공통 스타일을 써보고 싶네요 CSS에 너무 많은 에너지 소모를ㅠㅠ

jambottle replied:
`reset.scss`는 관습처럼 사용되는 스타일이라서 [이 페이지](https://meyerweb.com/eric/tools/css/reset/)에서 가져왔구요-
(다른 방식으로 reset.scss 작성하는 경우도 많다고 알고 있습니다!)

`common.scss`는 팀 내에서나 자체적으로 프로젝트 전반에 공통적으로 적용할 스타일을 설정해주는 파일이라서
제가 습관적으로 반복하는 스타일링에 한해서만 최소한으로 적용시켜주고 있어요. 통일감 있게 스타일링 할 때 좋아요 🤙

watertheqlant commented on mission/src/views/ItemInfo.vue:
```diff
@@ -1,24 +1,306 @@
+        margin-right: 2px;
+        font-size: 28px;
+        opacity: 0.4;
+        cursor: pointer;
```
cursor도 고려해서 작업하셨네요 
작은 부분도 놓치지 않는 모습 멋져요 :D

watertheqlant commented on mission/tests/unit/ItemInfo/ItemInfo.spec.js:
```diff
@@ -1,10 +1,60 @@
+  });
+
+  it('displays reviews with post and attatched photo of each review', () => {
+    expect(wrapper.get('b[data-test="review-name"]').exists()).toBe(true);
+    expect(wrapper.get('span[data-test="review-date"]').exists()).toBe(true);
+    expect(wrapper.get('h6[data-test="review-title"]').exists()).toBe(true);
+    expect(wrapper.get('p[data-test="review-content"]').exists()).toBe(true);
+    expect(wrapper.get('div[data-test="review-photo"]').exists()).toBe(true);
```
아..하! 이런식으로 한번에 묶어서 테스트를 할 수도 있군요. 
저는 exist 하나씩 it case를 만들어 주니, 중복된 코드가 너무 많아졌었는데, 
이렇게 하니 훨씬 가독성이 좋고 한눈에 들어오네요:)

kyubumjang replied:
저도 이부분에서 테스트 코드를 이런식으로 작성할 수도 있구나 하나 배워가요!! 가독성이 아주 좋은 것 같습니다.

jambottle committed:
Refactor: Mock Data 구조 수정 및 Inline Expression을 Computed Property로 대체
- itemData.js 내에서 관련성이 높은 상위 단계의 개념끼리 묶는 방향으로 데이터 구조를 수정
- template 내에서 데이터 가공을 위해 작성된 Inline Expression을 Computed Property로 대체

KyubumJang commented:
안녕하세요! 재원님!!
Code Presentation 잘봤습니다.
테스트 코드를 먼저 작성하고 실제로 구현하는 과정을 직접 거치시는 것이 기억에 남습니다. 항상 적용해보려 하다가 시간에 쫓겨 기능 구현에 급급하는 것 같네요..
<div> 태그만을 사용하지 않고 페이지를 구현할 수 있는 방법에 대해서도 알아갈 수 있었습니다.
항상 리뷰 정성스럽게 남겨주셔서 감사합니다! 이번 주 고생 많으셨습니다. 다음 주도 화이팅입니다!!

kyubumjang commented on mission/tests/unit/ItemInfo/ItemInfo.spec.js:
```diff
@@ -1,10 +1,40 @@
+
+  it('shows fixed footer with a button which contains discount price', () => {
+    expect(wrapper.get('footer[data-test="footer-wrapper"]').exists()).toBe(true);
+    expect(wrapper.get('button[data-test="footer-button"]').exists()).toBe(true);
```
테스트 코드를 테스트 할 때 data-test로 다 태그를 달아서 할 수 있었군요. 테스트 코드 작성해서 항상 어려움을 겪는 것 같아요. 저번에 재원님이 주신 유투브 영상을 아직 못봤는데 그 부분이랑 강사님께서 주신 파일 잘 읽어보고 작성하는 연습을 해야겠습니다. 잘 배워갑니다!

jambottle committed:
Refactor: Mock Data 구조 수정 및 ItemInfo 컴포넌트에 수정사항 반영
- reviewData.js 내에서 관련성이 높은 상위 단계의 개념끼리 묶는 방향으로 데이터 구조를 수정
- itemData.js 내에서도 D/B 측면의 관점을 고려하여 shop, item 항목마다 id 값을 추가적으로 부여
- 각 파일별로 테스트 목적에서 데이터를 하나씩 추가하고, 수정사항들을 실제 뷰에 반영하여 확인

jambottle committed:
Refactor: ItemInfo.spec.js 테스트 코드 1차 리팩토링
- setData 메서드를 활용해 Data Property에 관한 테스트 코드를 전반적으로 추가

jambottle committed:
Modify: ItemInfo 컴포넌트로부터 Shop/Review 컴포넌트 분리
- Item 데이터와 직접적인 관련성이 떨어지는 부분을 별도의 하위 컴포넌트로 분리
- 하위 컴포넌트를 ItemInfo에 등록하고, props로 필요한 데이터만 선택적으로 전달
- emit을 활용해 Shop 컴포넌트에서 상위 컴포넌트의 toggleLike 이벤트를 실행

jambottle committed:
Refactor: ItemInfo.spec.js 테스트 코드 2차 리팩토링
- 직접적인 관련성이 높은 데이터 세트를 기준으로 테스트 코드를 세 그룹으로 구분 (Shop/Item/Review)
- 같은 데이터를 공유하는 테스트 그룹의 경우, beforeEach 메서드를 활용해 테스트마다 데이터 초기화

jambottle commented:
**📌 테스트 코드 관련 리팩토링 관련**

> 지난 주에는 DOM Node에 표시되는 텍스트 값을 단순 확인하기 위해 `setValue 메서드`를 활용했다면, 이번 주에는 의도한 대로 렌더링이 성공했는지의 여부를 검증하기 위해 `setData 메서드`를 활용해보았습니다. **TDD 방법론과 E2E Testing을 적절히 조합하면, 좋은 시너지 효과를 낼 수 있겠다**는 생각이 화면을 구현하면서 들었습니다.

- 직접적인 관련성이 높은 데이터 세트를 기준으로 테스트 코드를 세 그룹으로 구분 (Shop/Item/Review)
- 같은 데이터를 공유하는 테스트 그룹의 경우, beforeEach 메서드를 활용해 테스트마다 데이터 초기화

**📌 느낀 점과 추가로 알게 된 점 관련**

> 다른 분들이 화면 구현 전에 와이어 프레임이나 데이터 명세서를 먼저 작성하신 사례를 보면서 **사전 계획 단계의 중요성**을 거듭 깨달았습니다!
저 역시도 머릿속으로만 계획하는 것이 아니라 **팀원들과의 명확한 소통을 위한 사전 작업 방식**에 대해 고민해볼 수 있는 계기가 되었습니다.

- 아무리 _"Mock"_ Data일지라도 데이터 구조를 짤 때, 충분히 체계적이고 세부적이어야 화면 구현할 때에도 도움 된다는 것을 인지
- 데이터 가공을 위해 작성된 Inline Expression도 최대한 Computed Property로 대체하여 문서를 깔끔하게 유지할 것을 확인
